torch/distributed/pipelining/schedules.py

```diff
@@ -2735,6 +2736,257 @@ def _calculate_single_rank_operations(self, rank) -> list[Optional[_Action]]:
+        self.pipeline_order: dict[int, list[Optional[_Action]]] = {}
+        for rank in range(self.pp_group_size):
+            rank_ops = self._calculate_single_rank_operations(rank)
+            self.pipeline_order[rank] = rank_ops
+
+        # Initialize the pipeline order with communication necessary to run with _PipelineScheduleRuntime
+        self._load_actions(self.pipeline_order)
+
+    def _calculate_single_rank_operations(self, rank) -> list[Optional[_Action]]:
+        actions: list[Optional[_Action]] = []
+        counters: dict[
+            tuple[int, _ComputationType], int
+        ] = {}  # (stage_index, computation_type) -> mb_index
+        weight_queue = []  # Queue of (stage_index, mb_index) for pending weight actions
+
+        num_ranks = self.pp_group_size
+        num_chunks = self._n_microbatches
+
+        rank_to_stages = generate_rank_to_stage_mapping(
+            num_ranks, num_ranks * 2, style="v"
+        )
+        stage0_index, stage1_index = rank_to_stages[rank]
+
+        def add_action(
+            actions: list,
+            stage_index: int,
+            computation_type: _ComputationType,
+            counters: dict,
+            forward_stage: Optional[int] = None,
+            backward_stage: Optional[int] = None,
+        ):
+            """Helper method to add an action with automatic microbatch index tracking."""
+            if computation_type == _ComputationType.OVERLAP_F_B:
+                assert forward_stage is not None and backward_stage is not None
+                # Create overlapped forward+backward action with sub_actions
+                forward_key = (forward_stage, _ComputationType.FORWARD)
+                backward_key = (backward_stage, _ComputationType.FULL_BACKWARD)
+
+                forward_mb = counters.get(forward_key, 0)
+                backward_mb = counters.get(backward_key, 0)
+
+                sub_actions = (
+                    _Action(forward_stage, _ComputationType.FORWARD, forward_mb),
+                    _Action(
+                        backward_stage, _ComputationType.FULL_BACKWARD, backward_mb
+                    ),
+                )
+
+                # Update counters for sub_actions
+                counters[forward_key] = forward_mb + 1
+                counters[backward_key] = backward_mb + 1
+
+                # Handle FULL_BACKWARD counter updates
+                input_key = (backward_stage, _ComputationType.BACKWARD_INPUT)
+                weight_key = (backward_stage, _ComputationType.BACKWARD_WEIGHT)
+                counters[input_key] = counters.get(input_key, 0) + 1
+                counters[weight_key] = counters.get(weight_key, 0) + 1
+
+                # Add the overlapped action
+                actions.append(
+                    _Action(-1, _ComputationType.OVERLAP_F_B, None, sub_actions)
+                )
+            else:
+                # Regular single action
+                key = (stage_index, computation_type)
```

Comment:
so if you'd done one O_F_B and then you are doing your first dI, you'd want counters to report '1' so you know you're doing the 1st backward.  seems correct